### ISSUE TITLE
The `primitive` haskell lib does not build

### ISSUE THREAD

**Issue by ElvishJerricco** [WebGHC/ghc#12]
Seems we still have a codegen problem.

```
$ nix-build -A nixpkgsWasm.haskell.packages.ghcHEAD.primitive

@nix { "action": "setPhase", "phase": "setupCompilerEnvironmentPhase" }
setupCompilerEnvironmentPhase
Build with /nix/store/sr37s1m1c1wjlfbw3fbz2m4xy5bnk8d8-wasm32-unknown-unknown-wasm-ghc-8.5.20180424.
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/jn648vh2xy6iqxph86b6nryzjrlpzc2l-primitive-0.6.4.0.tar.gz
source root is primitive-0.6.4.0
setting SOURCE_DATE_EPOCH to timestamp 1527705516 of file primitive-0.6.4.0/test/primitive-tests.cabal
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
Replace Cabal file with edited version from http://hackage.haskell.org/package/primitive-0.6.4.0/revision/1.cabal.
Run jailbreak-cabal to lift version restrictions on build inputs.
@nix { "action": "setPhase", "phase": "compileBuildDriverPhase" }
compileBuildDriverPhase
setupCompileFlags: -package-db=/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/setup-package.conf.d -j4 -threaded
[1 of 1] Compiling Main             ( Setup.hs, /tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/Main.o )
Linking Setup ...
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
configureFlags: --verbose --prefix=/nix/store/bw7qmsj0r0pw4pi7k2rm8362gh3g6218-primitive-0.6.4.0-wasm32-unknown-unknown-wasm --libdir=$prefix/lib/$compiler --libsubdir=$pkgid --with-gcc=wasm32-unknown-unknown-wasm-clang --package-db=/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/package.conf.d --ghc-option=-j4 --disable-split-objs --enable-library-profiling --profiling-detail=all-functions --disable-profiling --disable-shared --disable-coverage --disable-executable-dynamic --disable-tests --enable-library-vanilla --enable-library-for-ghci --configure-option=--host=wasm32-unknown-unknown-wasm --with-ghc=wasm32-unknown-unknown-wasm-ghc --with-ghc-pkg=wasm32-unknown-unknown-wasm-ghc-pkg --with-gcc=wasm32-unknown-unknown-wasm-cc --with-ld=wasm32-unknown-unknown-wasm-ld --with-hsc2hs=wasm32-unknown-unknown-wasm-hsc2hs --with-strip=wasm32-unknown-unknown-wasm-strip --hsc2hs-option=--cross-compile --ghc-option=-optl --ghc-option=-Wl,--export=main --extra-include-dirs=/nix/store/ykykk9gjk2h6za56hzl2d4ys9q6wh15c-libiconv-1.15-wasm32-unknown-unknown-wasm/include --extra-lib-dirs=/nix/store/ykykk9gjk2h6za56hzl2d4ys9q6wh15c-libiconv-1.15-wasm32-unknown-unknown-wasm/lib
Using Parsec parser
Configuring primitive-0.6.4.0...
Dependency base -any: using base-4.12.0.0
Dependency ghc-prim -any: using ghc-prim-0.5.2.0
Dependency transformers -any: using transformers-0.5.5.0
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
Source component graph: component lib
Configured component graph:
    component primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7
        include base-4.12.0.0
        include ghc-prim-0.5.2.0
        include transformers-0.5.5.0
Linked component graph:
    unit primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7
        include base-4.12.0.0
        include ghc-prim-0.5.2.0
        include transformers-0.5.5.0
        Control.Monad.Primitive=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Control.Monad.Primitive,Data.Primitive=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Data.Primitive,Data.Primitive.Addr=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Data.Primitive.Addr,Data.Primitive.Array=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Data.Primitive.Array,Data.Primitive.ByteArray=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Data.Primitive.ByteArray,Data.Primitive.MVar=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Data.Primitive.MVar,Data.Primitive.MachDeps=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Data.Primitive.MachDeps,Data.Primitive.MutVar=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Data.Primitive.MutVar,Data.Primitive.PrimArray=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Data.Primitive.PrimArray,Data.Primitive.Ptr=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Data.Primitive.Ptr,Data.Primitive.SmallArray=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Data.Primitive.SmallArray,Data.Primitive.Types=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Data.Primitive.Types,Data.Primitive.UnliftedArray=primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7:Data.Primitive.UnliftedArray
Ready component graph:
    definite primitive-0.6.4.0-2YnD1Z7OxvBF6KH7JR7JY7
        depends base-4.12.0.0
        depends ghc-prim-0.5.2.0
        depends transformers-0.5.5.0
Using Cabal-2.1.0.0 compiled by ghc-8.5
Using compiler: ghc-8.5.20180424
Using install prefix:
/nix/store/bw7qmsj0r0pw4pi7k2rm8362gh3g6218-primitive-0.6.4.0-wasm32-unknown-unknown-wasm
Executables installed in:
/nix/store/bw7qmsj0r0pw4pi7k2rm8362gh3g6218-primitive-0.6.4.0-wasm32-unknown-unknown-wasm/bin
Libraries installed in:
/nix/store/bw7qmsj0r0pw4pi7k2rm8362gh3g6218-primitive-0.6.4.0-wasm32-unknown-unknown-wasm/lib/ghc-8.5.20180424/primitive-0.6.4.0
Dynamic Libraries installed in:
/nix/store/bw7qmsj0r0pw4pi7k2rm8362gh3g6218-primitive-0.6.4.0-wasm32-unknown-unknown-wasm/lib/ghc-8.5.20180424/wasm32-unknown-wasm-ghc-8.5.20180424
Private executables installed in:
/nix/store/bw7qmsj0r0pw4pi7k2rm8362gh3g6218-primitive-0.6.4.0-wasm32-unknown-unknown-wasm/libexec/wasm32-unknown-wasm-ghc-8.5.20180424/primitive-0.6.4.0
Data files installed in:
/nix/store/bw7qmsj0r0pw4pi7k2rm8362gh3g6218-primitive-0.6.4.0-wasm32-unknown-unknown-wasm/share/wasm32-unknown-wasm-ghc-8.5.20180424/primitive-0.6.4.0
Documentation installed in:
/nix/store/bw7qmsj0r0pw4pi7k2rm8362gh3g6218-primitive-0.6.4.0-wasm32-unknown-unknown-wasm/share/doc/wasm32-unknown-wasm-ghc-8.5.20180424/primitive-0.6.4.0
Configuration files installed in:
/nix/store/bw7qmsj0r0pw4pi7k2rm8362gh3g6218-primitive-0.6.4.0-wasm32-unknown-unknown-wasm/etc
No alex found
Using ar found on system at:
/nix/store/8da6py0i79w7qhx0zicqyd7icm1kqdq2-llvm-binutils-7.0.0/bin/wasm32-unknown-unknown-wasm-ar
No c2hs found
No cpphs found
No doctest found
Using gcc version 4.2.1 given by user at:
/nix/store/y8ykrnqgq04a9d08niyd8zfcpvi5hgzf-wasm32-unknown-unknown-wasm-clang-cross-wrapper/bin/wasm32-unknown-unknown-wasm-cc
Using ghc version 8.5.20180424 given by user at:
/nix/store/sr37s1m1c1wjlfbw3fbz2m4xy5bnk8d8-wasm32-unknown-unknown-wasm-ghc-8.5.20180424/bin/wasm32-unknown-unknown-wasm-ghc
Using ghc-pkg version 8.5.20180424 given by user at:
/nix/store/sr37s1m1c1wjlfbw3fbz2m4xy5bnk8d8-wasm32-unknown-unknown-wasm-ghc-8.5.20180424/bin/wasm32-unknown-unknown-wasm-ghc-pkg
No ghcjs found
No ghcjs-pkg found
No greencard found
Using haddock version 2.18.2 found on system at:
/nix/store/cvsnpd2jj0395k08vwidfjbqxv64586g-ghc-8.5.20180118/bin/haddock
No happy found
Using haskell-suite found on system at: haskell-suite-dummy-location
Using haskell-suite-pkg found on system at: haskell-suite-pkg-dummy-location
No hmake found
Using hpc version 0.67 found on system at:
/nix/store/cvsnpd2jj0395k08vwidfjbqxv64586g-ghc-8.5.20180118/bin/hpc
Using hsc2hs version 0.68.3 given by user at:
/nix/store/sr37s1m1c1wjlfbw3fbz2m4xy5bnk8d8-wasm32-unknown-unknown-wasm-ghc-8.5.20180424/bin/wasm32-unknown-unknown-wasm-hsc2hs
Using hscolour version 1.24 found on system at:
/nix/store/q0cprnyqwwlfmvqaamd1l4fka2h0xp0k-hscolour-1.24.4/bin/HsColour
No jhc found
Using ld given by user at:
/nix/store/y8ykrnqgq04a9d08niyd8zfcpvi5hgzf-wasm32-unknown-unknown-wasm-clang-cross-wrapper/bin/wasm32-unknown-unknown-wasm-ld
No lhc found
No lhc-pkg found
No pkg-config found
Using runghc version 8.5.20180118 found on system at:
/nix/store/cvsnpd2jj0395k08vwidfjbqxv64586g-ghc-8.5.20180118/bin/runghc
Using strip given by user at:
/nix/store/8da6py0i79w7qhx0zicqyd7icm1kqdq2-llvm-binutils-7.0.0/bin/wasm32-unknown-unknown-wasm-strip
Using tar found on system at:
/nix/store/wpdkxjpwj0yihaixbmamdlzrpw5vn944-gnutar-1.30/bin/tar
No uhc found
@nix { "action": "setPhase", "phase": "buildPhase" }
building
Preprocessing library for primitive-0.6.4.0..
Building library for primitive-0.6.4.0..
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
[ 1 of 15] Compiling Control.Monad.Primitive ( Control/Monad/Primitive.hs, dist/build/Control/Monad/Primitive.o )
clang-7: warning: argument unused during compilation: '-fno-strict-overflow' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
[ 2 of 15] Compiling Data.Primitive.Internal.Compat ( Data/Primitive/Internal/Compat.hs, dist/build/Data/Primitive/Internal/Compat.o )
clang-7: warning: argument unused during compilation: '-fno-strict-overflow' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
[ 3 of 15] Compiling Data.Primitive.Array ( Data/Primitive/Array.hs, dist/build/Data/Primitive/Array.o )
clang-7: warning: argument unused during compilation: '-fno-strict-overflow' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
[ 4 of 15] Compiling Data.Primitive.MVar ( Data/Primitive/MVar.hs, dist/build/Data/Primitive/MVar.o )
clang-7: warning: argument unused during compilation: '-fno-strict-overflow' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
[ 5 of 15] Compiling Data.Primitive.MachDeps ( Data/Primitive/MachDeps.hs, dist/build/Data/Primitive/MachDeps.o )
clang-7: warning: argument unused during compilation: '-fno-strict-overflow' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
[ 6 of 15] Compiling Data.Primitive.Internal.Operations ( Data/Primitive/Internal/Operations.hs, dist/build/Data/Primitive/Internal/Operations.o )
clang-7: warning: argument unused during compilation: '-fno-strict-overflow' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-nostdlib++' [-Wunused-command-line-argument]
clang-7: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:652:15: error:
     error: conflicting types for 'hsprimitive_memset_Word8'
    |
652 | {extern void (hsprimitive_memset_Word8)(void *, I_, W_, I_);
    |               ^
{extern void (hsprimitive_memset_Word8)(void *, I_, W_, I_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:97:15: error:
     note: previous declaration is here
   |
97 | {extern void (hsprimitive_memset_Word8)(void *, I_, W_, W_);
   |               ^
{extern void (hsprimitive_memset_Word8)(void *, I_, W_, W_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:763:15: error:
     error: conflicting types for 'hsprimitive_memset_Word16'
    |
763 | {extern void (hsprimitive_memset_Word16)(void *, I_, W_, I_);
    |               ^
{extern void (hsprimitive_memset_Word16)(void *, I_, W_, I_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:208:15: error:
     note: previous declaration is here
    |
208 | {extern void (hsprimitive_memset_Word16)(void *, I_, W_, W_);
    |               ^
{extern void (hsprimitive_memset_Word16)(void *, I_, W_, W_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:874:15: error:
     error: conflicting types for 'hsprimitive_memset_Word32'
    |
874 | {extern void (hsprimitive_memset_Word32)(void *, I_, W_, I_);
    |               ^
{extern void (hsprimitive_memset_Word32)(void *, I_, W_, I_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:319:15: error:
     note: previous declaration is here
    |
319 | {extern void (hsprimitive_memset_Word32)(void *, I_, W_, W_);
    |               ^
{extern void (hsprimitive_memset_Word32)(void *, I_, W_, W_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:985:15: error:
     error: conflicting types for 'hsprimitive_memset_Word64'
    |
985 | {extern void (hsprimitive_memset_Word64)(void *, I_, W_, StgInt64);
    |               ^
{extern void (hsprimitive_memset_Word64)(void *, I_, W_, StgInt64);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:430:15: error:
     note: previous declaration is here
    |
430 | {extern void (hsprimitive_memset_Word64)(void *, I_, W_, StgWord64);
    |               ^
{extern void (hsprimitive_memset_Word64)(void *, I_, W_, StgWord64);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:1096:15: error:
     error: conflicting types for 'hsprimitive_memset_Word'
     |
1096 | {extern void (hsprimitive_memset_Word)(void *, I_, W_, I_);
     |               ^
{extern void (hsprimitive_memset_Word)(void *, I_, W_, I_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:541:15: error:
     note: previous declaration is here
    |
541 | {extern void (hsprimitive_memset_Word)(void *, I_, W_, W_);
    |               ^
{extern void (hsprimitive_memset_Word)(void *, I_, W_, W_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:2206:15: error:
     error: conflicting types for 'hsprimitive_memset_Word8'
     |
2206 | {extern void (hsprimitive_memset_Word8)(void *, I_, W_, I_);
     |               ^
{extern void (hsprimitive_memset_Word8)(void *, I_, W_, I_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:97:15: error:
     note: previous declaration is here
   |
97 | {extern void (hsprimitive_memset_Word8)(void *, I_, W_, W_);
   |               ^
{extern void (hsprimitive_memset_Word8)(void *, I_, W_, W_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:2317:15: error:
     error: conflicting types for 'hsprimitive_memset_Word16'
     |
2317 | {extern void (hsprimitive_memset_Word16)(void *, I_, W_, I_);
     |               ^
{extern void (hsprimitive_memset_Word16)(void *, I_, W_, I_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:208:15: error:
     note: previous declaration is here
    |
208 | {extern void (hsprimitive_memset_Word16)(void *, I_, W_, W_);
    |               ^
{extern void (hsprimitive_memset_Word16)(void *, I_, W_, W_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:2428:15: error:
     error: conflicting types for 'hsprimitive_memset_Word32'
     |
2428 | {extern void (hsprimitive_memset_Word32)(void *, I_, W_, I_);
     |               ^
{extern void (hsprimitive_memset_Word32)(void *, I_, W_, I_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:319:15: error:
     note: previous declaration is here
    |
319 | {extern void (hsprimitive_memset_Word32)(void *, I_, W_, W_);
    |               ^
{extern void (hsprimitive_memset_Word32)(void *, I_, W_, W_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:2539:15: error:
     error: conflicting types for 'hsprimitive_memset_Word64'
     |
2539 | {extern void (hsprimitive_memset_Word64)(void *, I_, W_, StgInt64);
     |               ^
{extern void (hsprimitive_memset_Word64)(void *, I_, W_, StgInt64);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:430:15: error:
     note: previous declaration is here
    |
430 | {extern void (hsprimitive_memset_Word64)(void *, I_, W_, StgWord64);
    |               ^
{extern void (hsprimitive_memset_Word64)(void *, I_, W_, StgWord64);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:2650:15: error:
     error: conflicting types for 'hsprimitive_memset_Word'
     |
2650 | {extern void (hsprimitive_memset_Word)(void *, I_, W_, I_);
     |               ^
{extern void (hsprimitive_memset_Word)(void *, I_, W_, I_);
              ^

/tmp/nix-build-primitive-0.6.4.0-wasm32-unknown-unknown-wasm.drv-0/ghc128545_0/ghc_46.hc:541:15: error:
     note: previous declaration is here
    |
541 | {extern void (hsprimitive_memset_Word)(void *, I_, W_, W_);
    |               ^
{extern void (hsprimitive_memset_Word)(void *, I_, W_, W_);
              ^
10 errors generated.
`wasm32-unknown-unknown-wasm-cc' failed in phase `C Compiler'. (Exit code: 1)
```

**Comment by ElvishJerricco** [comment]
The problem appears to be these four lines.

```
foreign import ccall unsafe "primitive-memops.h hsprimitive_memset_Word"
  setWordArray# :: MutableByteArray# s -> CPtrdiff -> CSize -> Word# -> IO ()
foreign import ccall unsafe "primitive-memops.h hsprimitive_memset_Word"
  setIntArray# :: MutableByteArray# s -> CPtrdiff -> CSize -> Int# -> IO ()
foreign import ccall unsafe "primitive-memops.h hsprimitive_memset_Word"
  setWordOffAddr# :: Addr# -> CPtrdiff -> CSize -> Word# -> IO ()
foreign import ccall unsafe "primitive-memops.h hsprimitive_memset_Word"
  setIntOffAddr# :: Addr# -> CPtrdiff -> CSize -> Int# -> IO ()
```

These four foreign imports are importing the same symbol at different types. I could argue that this *should* break, because the same symbol only ever has one type. But the types do have the same runtime representation, so maybe it should be ok... Regardless, Clang isn't going to like it.

**Comment by ElvishJerricco** [comment]
I tried to fix this by replacing `ccall` with `capi`, but got a GHC panic:

```
ghc: panic! (the 'impossible' happened)
  (GHC version 8.2.2 for x86_64-unknown-linux):
	toCType
  Word#
  Call stack:
      CallStack (from HasCallStack):
        prettyCurrentCallStack, called at compiler/utils/Outputable.hs:1133:58 in ghc:Outputable
        callStackDoc, called at compiler/utils/Outputable.hs:1137:37 in ghc:Outputable
        pprPanic, called at compiler/deSugar/DsForeign.hs:726:17 in ghc:DsForeign
```

This panic appears to occur with an official native GHC release, so it's not at all specific to WebGHC.

**Comment by ElvishJerricco** [comment]
https://ghc.haskell.org/trac/ghc/ticket/14624

**Comment by ElvishJerricco** [comment]
Fixed in 35a703dad585639021eb88acdaf117837078eb47, though `primitive` still needs patching to use `capi`. Will push such a patch to wasm-cross soon.
